packages/puppeteer-core/src/common/util.ts

```diff
@@ -488,3 +489,65 @@ export function stringifyFunction(expression: Function): string {
   }
   return functionText;
 }
+
+/**
+ * @internal
+ */
+export async function* deepChildren(
+  handle: ElementHandle<Node>
+): AwaitableIterable<ElementHandle<Node>> {
+  const walker = await handle.evaluateHandle(element => {
+    return document.createTreeWalker(element, NodeFilter.SHOW_ELEMENT);
+  });
+  while (true) {
+    const handle = await walker.evaluateHandle(walker => {
+      const node = walker.nextNode();
+      if (!(node instanceof Element)) {
+        return;
+      }
+      return node.shadowRoot ?? node;
+    });
+    if (handle instanceof ElementHandle<Node>) {
+      yield handle;
+    } else {
+      break;
+    }
+  }
+  await walker.dispose();
+}
+
+/**
+ * @internal
+ */
+export async function* deepDescendents(
+  handle: ElementHandle<Node>
+): AwaitableIterable<ElementHandle<Node>> {
+  const walkers = await handle.evaluateHandle(element => {
+    return [document.createTreeWalker(element, NodeFilter.SHOW_ELEMENT)];
+  });
+  while (true) {
+    const handle = await walkers.evaluateHandle(walkers => {
```

Comment:
does it make the number of cdp calls to equal the number of shadow roots under an element? 

I don't think it's very efficient. Could we move the entirety of the search into the eval function to make it atomic and not so network heavy?

Reply:
This function is called whenever someone uses a `>>>` combinator. The number of CDP calls whenever used is equal to the number of elements in the elements tree excluding or included descendent trees (depending on the `deep*` function).

For querying a single element (which is most of the time), this is very efficient. It's also efficient if someone doesn't use `>>>`.

We should keep in mind that querying a shadow root is a very powerful, but expensive task. Since the largest use case is querying a single element, this is what we are optimizing for.

Reply:
[EDIT] With batching included now, the number of CDP calls is equal to the number of elements / batch size.